src/storage/compression/zstd.cpp

```diff
@@ -0,0 +1,1129 @@
+};
+
+//===--------------------------------------------------------------------===//
+// Analyze
+//===--------------------------------------------------------------------===//
+struct ZSTDAnalyzeState : public AnalyzeState {
+public:
+	ZSTDAnalyzeState(CompressionInfo &info, DBConfig &config)
+	    : AnalyzeState(info), config(config), compression_dict(nullptr), context(nullptr) {
+		context = duckdb_zstd::ZSTD_createCCtx();
+	}
+	~ZSTDAnalyzeState() override {
+		duckdb_zstd::ZSTD_freeCCtx(context);
+		duckdb_zstd::ZSTD_freeCDict(compression_dict);
+	}
+
+public:
+	inline void AppendString(const string_t &str) {
+		auto string_size = str.GetSize();
+		total_size += string_size;
+	}
+	inline idx_t GetVectorCount() const {
+		idx_t vector_count = count / STANDARD_VECTOR_SIZE;
+		vector_count += (count % STANDARD_VECTOR_SIZE) != 0;
+		return vector_count;
+	}
+
+	inline idx_t GetVectorMetadataSize() const {
+		auto vector_count = GetVectorCount();
+
+		idx_t vector_metadata_size = 0;
+		vector_metadata_size += sizeof(page_id_t) * vector_count;
+
+		vector_metadata_size = AlignValue<idx_t, sizeof(page_offset_t)>(vector_metadata_size);
+		vector_metadata_size += sizeof(page_offset_t) * vector_count;
+
+		vector_metadata_size = AlignValue<idx_t, sizeof(uncompressed_size_t)>(vector_metadata_size);
+		vector_metadata_size += sizeof(uncompressed_size_t) * vector_count;
+
+		vector_metadata_size = AlignValue<idx_t, sizeof(compressed_size_t)>(vector_metadata_size);
+		vector_metadata_size += sizeof(compressed_size_t) * vector_count;
+		return vector_metadata_size;
+	}
+
+public:
+	DBConfig &config;
+	ZSTDSamplingState sampling_state;
+
+	//! The trained 'dictBuffer' (populated in FinalAnalyze)
+	DictBuffer dict;
+	duckdb_zstd::ZSTD_CDict *compression_dict;
+
+	duckdb_zstd::ZSTD_CCtx *context;
+	idx_t total_size = 0;
+	idx_t count = 0;
+};
+
+unique_ptr<AnalyzeState> ZSTDStorage::StringInitAnalyze(ColumnData &col_data, PhysicalType type) {
+	CompressionInfo info(col_data.GetBlockManager().GetBlockSize());
+	auto &data_table_info = col_data.info;
+	auto &attached_db = data_table_info.GetDB();
+	auto &config = DBConfig::Get(attached_db);
+
+	return make_uniq<ZSTDAnalyzeState>(info, config);
+}
+
+// Determines wether compression is possible and calculates sizes for the FinalAnalyze
+bool ZSTDStorage::StringAnalyze(AnalyzeState &state_p, Vector &input, idx_t count) {
+	auto &state = state_p.Cast<ZSTDAnalyzeState>();
+	UnifiedVectorFormat vdata;
+	input.ToUnifiedFormat(count, vdata);
+
+	auto data = UnifiedVectorFormat::GetData<string_t>(vdata);
+	for (idx_t i = 0; i < count; i++) {
+		auto idx = vdata.sel->get_index(i);
+		if (!vdata.validity.RowIsValid(idx)) {
+			continue;
+		}
+		auto &str = data[idx];
+		auto string_size = str.GetSize();
+		state.total_size += string_size;
+	}
+	state.sampling_state.Sample(input, count);
+	state.count += count;
+	return true;
+}
+
+DictBuffer CreateDictFromSamples(ZSTDAnalyzeState &state) {
```

Comment:
Sorry to just mingle here in this PR. Great work, indeed zstd is kinda hard to avoid, certainly for json (I do have some more concrete plans for enhancing FSST for columns with long strings with common suffixes, as URLs often have).

Anyway, I imagine the reason for using zstd with a trained dictionary is to have some way of to do random access into specific vectors of a rowgroup. But, using the same dictionary throughout a rowgroup could give significantly worse compression than plainly zstd-compression of all string data belonging to one vector, which gives similar random access possibility.. In my recollection, zstd starts having good effect already >30KB of data, and with 2K vectors, strings have just to be 15 bytes long for that, so it could work quite often.

Was a comparison done on the pro's and cons? I could also imagine two zstd schemes, one just compressing strings vector-at-a-time and one using a trained dictionary.

BTW we do have extensive real-world datasets lying around for testing, namely:
- https://homepages.cwi.nl/~boncz/PublicBIbenchmark/
- https://homepages.cwi.nl/~boncz/RealNest/ and
- https://homepages.cwi.nl/~boncz/NextiaJD/

Reply:
Thanks for mingling, always appreciate extra information on the subject 👍 

The thought process was more along the lines of: if we give ZSTD more context on the data we want to compress, it has the chance to great a better dictionary which improves compression ratio and decompression speed

We're saving where a vector starts to provide random access, but we also end the frame for each encoded vector so I believe this isn't seen as random access in ZSTD terms.

I was playing around with the idea of not writing a dictionary today, that would make the analyze step much simpler and cleanly solve the regression created by the expensive analyze method of this compression function.

I scrapped it due to not finding enough resources on what the trade offs are in terms of compression and decompression ratio + speed.
But I'm open to revisiting the idea!

Reply:
indeed the compression should be improved with a dictionary when strings are shorter.

One could do a compression rate experiment, by feeding in values from the mentioned datasets into a scripted implementation (not yet integrated in DuckDB) that measures compression ratio with and without a pre-trained dictionary. To learn what the trade-offs are. But indeed, that takes time.

And I now realize that likely compression ratio would not get much worse for larger strings with a dictionary (only storing the dictionary will eat into compression ratio). So you are right that no dictionary would have likely the main advantage that writing would get faster.

Reply:
I've been running some experiments with the datasets you provided and it does seem like the shared dictionary is not very relevant for compression ration, at most I've seen a 0.1~ compression ratio decrease (from `4.96~` to `4.87~`) and in other cases I've seen it actually improve the compression ratio by a non neglible amount (one instance of `1.7647059` to `3.109589`, using the `Arade_1.csv.bz2` dataset)

So I've decided to remove the shared dictionary, perhaps we can introduce a variant of the zstd compression for smaller strings where we reintroduce the shared dictionary

Reply:
The loss of compression ratio is likely explained by the limitations we're imposing on the amount of samples and the  total size of the sample buffer, vectors that don't fit the pattern will compress very badly.

This can probably be patched somewhat by having smarter logic around choosing samples, but there will always be edgecases.
That, combined with the cost of training the shared dictionary not being insignificant make me think this isn't worth doing in the general case

Reply:
Thanks for adding the benchmarks. 

Most of the columns you selected would benefit from an envisioned Local Common Prefix/Repeat optimized FSST. It looks like a MSc student will implement that. That means we would have that by spring 2025.

It would maybe also be good to test some JSON, people store that a lot, I guess. I expect it to compress much better with zstd than with FSST, and there the common prefix optimization will not help.